### DISCUSSION TITLE
How would I setup a cluster including a bare metal server?

### DISCUSSION THREAD

**Discussion by if-mar** [vitobotta/hetzner-k3s#489] [Q&A] [answered]
Hi, I'm currently taking a closer look at hetzner-k3s and have already created and used a few test clusters. For productive use, however, I would like to add a bare-metal worker node with more resources than the cloud instances offer to the cluster. 
I can reach the bare-metal server from the cloud network via a vSwitch.
How would I have to adapt the cluster-config yaml file to realise this? Thanks for any support!

**Comment by vitobotta** [comment]
Hey, currently, this feature isn’t supported out of the box, but I’m thinking about adding it in the future since it could be really useful.

For now, you can try to manually connect your dedicated server to the cluster. To do this, you need to have updated to `hetzner-k3s` version 2.0.9 (which is the latest one).

Here’s how you can do it:

1. First, clone the `hetzner-k3s` repository to your local machine and open it as a development container in VS Code. This is easier because it already has Crystal and other dependencies, and you’ll need Docker for this to work (see https://github.com/vitobotta/hetzner-k3s/blob/main/docs/Contributing_and_support.md#developing-with-vscode).
2. Then, open the `src/kubernetes/installer.cr` file and add the line `puts install_script` right after line 147.
3. After that, run the `create` command using `hetzner-k3s` like this:
```bash
crystal run src/hetzner-k3s.cr -- create --config <path to your cluster's yaml file>
```

The `create` command will not change anything in the cluster unless you have made changes to the configuration file. However, because of the line you added in step 2, it will print out the full command that `hetzner-k3s` uses to join a node to the cluster.

Lastly, run the join command on your dedicated server. If the dedicated server is in the same network as the cluster, it should work. Keep in mind that I haven’t tried this myself yet. Good luck, and let me know how it goes!

**Reply by if-mar** [reply]
Thank you for your instructions! I will try and keep you updated.

**Reply by ledniy** [reply]
@if-mar Could you please let us know if you succeeded?

**Reply by if-mar** [reply]
Unfortunately, I hadn't yet found the time to try it out as other projects had higher priority. If you test it, please let us know if it worked.

**Reply by ledniy** [reply]
Unfortunately it's not enough to run the commands from the install_script, after that the node has following taints:
```
│ Taints:             node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule                                                                                                       │
│                     node.kubernetes.io/network-unavailable:NoSchedule   
```

and such events:
```
│   Warning  FailedToCreateRoute       7m12s                  route_controller  Could not create route 1a015b70-7786-49fa-8b09-c9616a50cde1 10.244.6.0/24 for node bm-node-1 after 329.995 │
│ 066ms: hcloud/CreateRoute: hcops/AllServersCache.ByName: bm-node-1 hcops/AllServersCache.getCache: not found                                                                             │
│   Warning  FailedToCreateRoute       6m42s                  route_controller  Could not create route 1a015b70-7786-49fa-8b09-c9616a50cde1 10.244.6.0/24 for node bm-node-1 after 312.246 │
│ 744ms: hcloud/CreateRoute: hcops/AllServersCache.ByName: bm-node-1 hcops/AllServersCache.getCache: not found                                                                             │
│   Warning  FailedToCreateRoute       6m12s                  route_controller  Could not create route 1a015b70-7786-49fa-8b09-c9616a50cde1 10.244.6.0/24 for node bm-node-1 after 68.208� │
│ �s: hcloud/CreateRoute: hcops/AllServersCache.ByName: bm-node-1 hcops/AllServersCache.getCache: not found                                                                                │
│   Warning  FailedToCreateRoute       5m42s                  route_controller  Could not create route 1a015b70-7786-49fa-8b09-c9616a50cde1 10.244.6.0/24 for node bm-node-1 after 306.375 │
│ 643ms: hcloud/CreateRoute: hcops/AllServersCache.ByName: bm-node-1 hcops/AllServersCache.getCache: not found                                                                             │
│   Warning  FailedToCreateRoute       5m12s                  route_controller  Could not create route 1a015b70-7786-49fa-8b09-c9616a50cde1 10.244.6.0/24 for node bm-node-1 after 48.912� │
│ �s: hcloud/CreateRoute: hcops/AllServersCache.ByName: bm-node-1 hcops/AllServersCache.getCache: not found                                                                                │
│   Warning  FailedToCreateRoute       4m42s                  route_controller  Could not create route 1a015b70-7786-49fa-8b09-c9616a50cde1 10.244.6.0/24 for node bm-node-1 after 316.188 │
│ 852ms: hcloud/CreateRoute: hcops/AllServersCache.ByName: bm-node-1 hcops/AllServersCache.getCache: not found                                                                             │
│   Warning  FailedToCreateRoute       4m12s                  route_controller  Could not create route 1a015b70-7786-49fa-8b09-c9616a50cde1 10.244.6.0/24 for node bm-node-1 after 104.155 │
│ µs: hcloud/CreateRoute: hcops/AllServersCache.ByName: bm-node-1 hcops/AllServersCache.getCache: not found                                                                                │
│   Warning  FailedToCreateRoute       3m42s                  route_controller  Could not create route 1a015b70-7786-49fa-8b09-c9616a50cde1 10.244.6.0/24 for node bm-node-1 after 290.382 │
│ 223ms: hcloud/CreateRoute: hcops/AllServersCache.ByName: bm-node-1 hcops/AllServersCache.getCache: not found                                                                             │
│   Warning  FailedToCreateRoute       3m12s                  route_controller  Could not create route 1a015b70-7786-49fa-8b09-c9616a50cde1 10.244.6.0/24 for node bm-node-1 after 299.132 │
│ 136ms: hcloud/CreateRoute: hcops/AllServersCache.ByName: bm-node-1 hcops/AllServersCache.getCache: not found                                                                             │
│   Warning  FailedToCreateRoute       12s (x6 over 2m42s)    route_controller  (combined from similar events): Could not create route 1a015b70-7786-49fa-8b09-c9616a50cde1 10.244.6.0/24  │
│ for node bm-node-1 after 310.252463ms: hcloud/CreateRoute: hcops/AllServersCache.ByName: bm-node-1 hcops/AllServersCache.getCache: not found 
```

**Reply by if-mar** [reply]
@ledniy where you able to fix the mentioned problems?

**Reply by ledniy** [reply]
@if-mar nope, I gave up on hybrid cluster, and just using dedicated servers with k3s provisioned manually

**Reply by if-mar** [reply]
@vitobotta do you have any idea what fails here and could you guide us fixing?

**Reply by vitobotta** [reply]
I think I see the problem. The Hetzner Cloud Controller Manager is designed for Hetzner Cloud, so when a new node joins the cluster it needs to create the necessary routes for the new node, but that won't work of course with dedicated servers. I think the same problem might happen with the CSI driver.

The only workaround I can think of is to avoid the installation of the CCM and CSI drivers. I haven't tried it because I'd have to purchase dedicated servers just to test this, but perhaps you could try it? 

Check out the 2.1.1 branch and comment out these lines:

```crystal
    Kubernetes::Software::Hetzner::Secret.new(configuration, settings).create
    Kubernetes::Software::Hetzner::CloudControllerManager.new(configuration, settings).install
    Kubernetes::Software::Hetzner::CSIDriver.new(configuration, settings).install
```

I suspect that with this change it should be possible to add dedis to the cluster but it needs to be tested. If you can test this and it works, I can make the installation of those components optional (enabled by default).

Keep in mind that without CCM and CSI though, you cannot create services of type `LoadBalancer` nor persistent volumes using the Hetzner block storage, directly from the cluster. So load balancers and storage must be managed manually or with other means.

**Reply by vitobotta** [reply]
I am confused though because I think in the past some users have gotten the combo working.

**Reply by if-mar** [reply]
I have now had some time to try out the setup. 
### Dedicated Server
I bought a dedicated server and followed https://docs.hetzner.com/cloud/networks/connect-dedi-vswitch to connect it to a cloud network. The following lines where added to `/etc/network/interfaces`
```
auto enp0s31f6.4001
iface enp0s31f6.4001 inet static
  address 10.0.1.3
  netmask 255.255.255.0
  vlan-raw-device enp0s31f6
  mtu 1400
  up ip route add 10.0.0.0/16 via 10.0.1.1 dev enp0s31f6.4001
  down ip route del 10.0.0.0/16 via 10.0.1.1 dev enp0s31f6.4001
```
This allows me to ping the cloud server from the bare-metal server and vice versa.

### hetzner-k3s
To allow communication with the dedicated server, I manually created a cloud-network ("test-network") using 10.0.0.0/16 and two subnets:

- 10.0.0.0/24 -> connects the cloud servers
- 10.0.1.0/24 -> connects the dedicated server via vSwitch

This is my cluster-config.yaml:
```yaml
---
hetzner_token: <HCLOUD_TOKEN>
cluster_name: bm-test
kubeconfig_path: "./kubeconfig"
k3s_version: v1.30.3+k3s1

networking:
  ssh:
    port: 22
    use_agent: false # set to true if your key has a passphrase
    public_key_path: "~/.ssh/id_rsa.pub"
    private_key_path: "~/.ssh/id_rsa"
  allowed_networks:
    ssh:
      - 0.0.0.0/0
    api: # this will firewall port 6443 on the nodes
      - 0.0.0.0/0
  public_network:
    ipv4: true
    ipv6: true
  private_network:
    enabled: true
    # subnet: 10.0.0.0/16
    existing_network_name: "test-network"
  cni:
    enabled: true
    encryption: false
    mode: flannel

  # cluster_cidr: 10.244.0.0/16 # optional: a custom IPv4/IPv6 network CIDR to use for pod IPs
  # service_cidr: 10.43.0.0/16 # optional: a custom IPv4/IPv6 network CIDR to use for service IPs. Warning, if you change this, you should also change cluster_dns!
  # cluster_dns: 10.43.0.10 # optional: IPv4 Cluster IP for coredns service. Needs to be an address from the service_cidr range

datastore:
  mode: etcd # etcd (default) or external
  # external_datastore_endpoint: postgres://....

schedule_workloads_on_masters: false

# image: rocky-9 # optional: default is ubuntu-24.04
# autoscaling_image: 103908130 # optional, defaults to the `image` setting
# snapshot_os: microos # optional: specified the os type when using a custom snapshot

masters_pool:
  instance_type: cx22
  instance_count: 1
  location: fsn1

worker_node_pools:
- name: small-static
  instance_type: cx32
  instance_count: 1
  location: fsn1
- name: medium-autoscaled
  instance_type: cpx31
  instance_count: 2
  location: fsn1
  autoscaling:
    enabled: true
    min_instances: 0
    max_instances: 3

embedded_registry_mirror:
  enabled: false # Check if your k3s version is compatible before enabling this option. You can find more information at https://docs.k3s.io/installation/registry-mirror

api_server_hostname: cluster.mydomain.com # optional: DNS for the k8s API LoadBalancer. After the script has run, create a DNS record with the address of the API LoadBalancer.
```

I'm using the current version (v2.2.7) of hetzner-k3s and added  the line `log_line '...this is the worker install script: #{worker_install_script(master_count)}', log_prefix: "Instance #{worker.name}"`  in the function `deploy_k3s_to_worker` in `src/kubernetes/installer.cr`. This prints me the used install script when running the create command.

I had to slightly adjust the install script a bit so that it would run, but in the end k3s is installed. However, I get the following log messages (journalctl -xeu k3s-agent.service):
```
Mar 04 14:33:20 Debian-bookworm-latest-amd64-base k3s[305466]: time="2025-03-04T14:33:20+01:00" level=info msg="Network policy controller waiting for removal of node.cloudprovider.kubernetes.io/uninitialized taint"
Mar 04 14:33:22 Debian-bookworm-latest-amd64-base k3s[305466]: time="2025-03-04T14:33:22+01:00" level=info msg="Network policy controller waiting for removal of node.cloudprovider.kubernetes.io/uninitialized taint"
Mar 04 14:33:24 Debian-bookworm-latest-amd64-base k3s[305466]: time="2025-03-04T14:33:24+01:00" level=info msg="Network policy controller waiting for removal of node.cloudprovider.kubernetes.io/uninitialized taint"
Mar 04 14:33:26 Debian-bookworm-latest-amd64-base k3s[305466]: time="2025-03-04T14:33:26+01:00" level=info msg="Network policy controller waiting for removal of node.cloudprovider.kubernetes.io/uninitialized taint"
Mar 04 14:33:28 Debian-bookworm-latest-amd64-base k3s[305466]: time="2025-03-04T14:33:28+01:00" level=info msg="Network policy controller waiting for removal of node.cloudprovider.kubernetes.io/uninitialized taint"
Mar 04 14:33:30 Debian-bookworm-latest-amd64-base k3s[305466]: time="2025-03-04T14:33:30+01:00" level=info msg="Network policy controller waiting for removal of node.cloudprovider.kubernetes.io/uninitialized taint"
Mar 04 14:33:32 Debian-bookworm-latest-amd64-base k3s[305466]: time="2025-03-04T14:33:32+01:00" level=info msg="Network policy controller waiting for removal of node.cloudprovider.kubernetes.io/uninitialized taint"
Mar 04 14:33:34 Debian-bookworm-latest-amd64-base k3s[305466]: time="2025-03-04T14:33:34+01:00" level=info msg="Network policy controller waiting for removal of node.cloudprovider.kubernetes.io/uninitialized taint"
Mar 04 14:33:35 Debian-bookworm-latest-amd64-base k3s[305466]: I0304 14:33:35.040524  305466 scope.go:117] "RemoveContainer" containerID="b523c0ae214a9637f516f80cdf59618b45f32b37b828dea334a6eca76c5d0db7"
Mar 04 14:33:35 Debian-bookworm-latest-amd64-base k3s[305466]: I0304 14:33:35.040760  305466 scope.go:117] "RemoveContainer" containerID="a123465af5a63887773e380bd5f905c36f9a14bc6ad146eb1f2b4b634894e4d9"
Mar 04 14:33:35 Debian-bookworm-latest-amd64-base k3s[305466]: E0304 14:33:35.041688  305466 pod_workers.go:1298] "Error syncing pod, skipping" err="[failed to \"StartContainer\" for \"csi-node-driver-registrar\" with CrashLoopBackOff: \"back-off 5m0s restarting failed>
M
```

When I run `kubectl get nodes --all-namespaces -o wide` from the DEV-Container, I get the output 
```
NAME                                STATUS   ROLES                       AGE    VERSION        INTERNAL-IP   EXTERNAL-IP       OS-IMAGE                         KERNEL-VERSION     CONTAINER-RUNTIME
bm-server                           Ready    <none>                      47m    v1.30.3+k3s1   10.0.1.3      <none>            Debian GNU/Linux 12 (bookworm)   6.1.0-31-amd64     containerd://1.7.17-k3s1
bm-test-master1                     Ready    control-plane,etcd,master   105m   v1.30.3+k3s1   10.0.0.2      xxx.xxx.xxx.xxx   Ubuntu 24.04.2 LTS               6.8.0-54-generic   containerd://1.7.17-k3s1
bm-test-pool-small-static-worker1   Ready    <none>                      104m   v1.30.3+k3s1   10.0.0.3      xxx.xxx.xxx.xxx   Ubuntu 24.04.2 LTS               6.8.0-54-generic   containerd://1.7.17-k3s1
```

The Hetzner Cloud Controller Manager supports dedicated servers since v1.19.0, so theoretically it should be possible as far I understand?

Maybe someone could guide me in the right direction. I would really need a node with more ressources in my cluster (which cloud servers not provide) to run my tasks.
Thanks for any help!

**Reply by vitobotta** [reply]
I haven't had the chance to try a cluster with dedicated servers myself yet. If you're up for it, you could reset the OS on your dedicated server and let me have temporary access. That way, I can check things out without having to buy a server. Once I'm done, you could just reset it again. What do you think?

**Reply by if-mar** [reply]
Hi @vitobotta, I would really appreciate this - thank you! I will send you the server details by email.

**Comment by ledniy** [comment]
@vitobotta Did you think about adding support for such nodes? It would be nice even if it will be half-manual, so we can have something like this:

```
- name: my-bare-metal-1
  instance_type: bare-metal
  ip: 1.1.1.1
  ```
So, to make it work we'll need to manually order dedicated server and do manual setup (install correct image, add vSwitch and correct config for it), after that I believe the tool can setup everything else?

**Reply by vitobotta** [reply]
Like I mentioned in my previous post in this thread, yes, adding support for dedicated servers is on my radar, but I don't have an ETA for now. But it's on my list

**Reply by simjak** [reply]
I was reading this https://community.hetzner.com/tutorials/kubernetes-on-hetzner-using-bare-metal-servers
It seems they somehow managed to plug dedicated nodes to Kubernetes cluster

**Reply by vitobotta** [reply]
@simjak

The ClusteAPI etc approach is a lot more complicated compared to what hetzner-k3s offers. Plus, the article seems to be talking about a cluster made of baremetal servers only, not one that mixes them with cloud instances like we're dealing with here.

I'm planning to add support for "external nodes" in hetzner-k3s. This would allow adding dedicated servers or even nodes from another provider if needed for any reason.

The only catch is that CCM and CSI can't be installed since they're designed to work with cloud instances only.

**Reply by neoromantique** [reply]
@vitobotta Would dedicated server only cluster be out of scope for the project?

**Reply by vitobotta** [reply]
@neoromantique For now, yes. Running dedicated servers has quite a few limits when compared to cloud instances-things like no API control, slow setup times, and so on. So even though I plan to support dedicated servers as nodes, the control plane will probably stay on the cloud for now. Using only dedicated servers would defeat most of the features and automation that hetzner-k3s provides. If you want to create a cluster with dedicated servers only, check out this project https://github.com/alexellis/k3sup

**Comment by vitobotta** [comment, marked as answer]
Thanks a lot to @if-mar for providing me access to a dedicated server as well as a Hetzner project with the necessary vSwitch setup, so that I could conduct some investigation.

I got a cluster with the dedicated server working nicely, for the most part. Unfortunately it seems that even the latest version of the HCCM doesn't handle dedicated servers properly as it's supposed to.

These are the steps I took to get the cluster working more or less correctly:

1 - Network setup with vSwitch link - done by @if-mar , who could add some instructions here as well
2 - Create the cluster with the cloud nodes as usual
3 - Create the file `/etc/k8s-resolv.conf` on the dedicated server with the single line `nameserver 8.8.8.8`
4 - Run the script at https://gist.github.com/vitobotta/68b13edfe242fbf5c229bb347d7b69f6 on the dedicated server, after replacing the token and the subnet (for vSwitch) with the correct values. You can find the correct token at `/var/lib/rancher/k3s/server/token` on a master; the script will hang for now since the node is still uninitialized according to the HCCM
5 - Edit the HCCM deployment and add the following environment variable:

```yaml
        - name: HCLOUD_NETWORK_ROUTES_ENABLED
          value: “false”
```
This disables the creation of routes by the HCCM but that should be fine because it's taken care by the CNI, unless you have some weird setup.
6 - Once the HCCM has restarted, run `kubectl taint nodes <node> node.cloudprovider.kubernetes.io/uninitialized-`, wait a few moments, then run `kubectl taint nodes <node> node.kubernetes.io/network-unavailable-`
7 - The node now should be initialized, and its private IP detected correctly, which is important for use with load balancers.
8 - Deployments etc should work fine. In the case a deployment requires a PVC with the `hcloud-volumes` storage class, the pod will be scheduled on a cloud node as expected.

Caveat:
- You need to remove the taints as per step 6 every time the HCCM restarts, otherwise new pods won't be scheduled on the dedicated server.

@if-mar Please try this and let me know how it goes. I will automate this stuff in hetzner-k3s when I have a little time. I am thinking of adding a new `add-robot-server` command which - given SSH access and the subnet for the vSwitch network - sets this up automatically.

**Reply by if-mar** [reply]
@vitobotta I can confirm that the node is added to the cluster. Sadly the taint `node.kubernetes.io/network-unavailable` is automatically added again as soon as I remove it. 
```shell
$ kubectl get nodes -o wide
NAME                                              STATUS     ROLES                       AGE    VERSION        INTERNAL-IP   EXTERNAL-IP      OS-IMAGE                         KERNEL-VERSION     CONTAINER-RUNTIME
debian-bookworm-latest-amd64-base                 Ready      <none>                      170m   v1.30.8+k3s1   10.0.1.2      <none>           Debian GNU/Linux 12 (bookworm)   6.1.0-28-amd64     containerd://1.7.23-k3s2
ifos-cluster-master1                              Ready      control-plane,etcd,master   3h     v1.30.3+k3s1   10.0.0.2      168.119.xxx.xxx   Ubuntu 24.04.2 LTS               6.8.0-54-generic   containerd://1.7.17-k3s1
ifos-cluster-master2                              Ready      control-plane,etcd,master   3h     v1.30.3+k3s1   10.0.0.3      138.199.xxx.xxx   Ubuntu 24.04.2 LTS               6.8.0-54-generic   containerd://1.7.17-k3s1
ifos-cluster-master3                              Ready      control-plane,etcd,master   179m   v1.30.3+k3s1   10.0.0.4      116.202.xxx.xxx   Ubuntu 24.04.2 LTS               6.8.0-54-generic   containerd://1.7.17-k3s1
```

The internal IP seems to be detected correctly, but external ip is `<none>`. Has this something to do with the taint?

**Reply by vitobotta** [reply]
Hi @if-mar , from what I've seen the taint is added again when restarting the HCCM. Does it happen otherwise for you? Re: external IP I am not sure. This is mostly a workaround to get some workloads running on the dedis, but without proper support from the HCCM it's hard to get it working perfectly.

**Reply by if-mar** [reply]
Hi @vitobotta, after serveral tries and some investigating, I managed to get it working. Here is, what I did:

1. In Hetzner Robot create a new vSwitch and add the root server.
2. On the root server, add the following lines to `/etc/network/interfaces` and adapt them to your needs. In this example the VLAN Id is `4001` and the internal IP address of the root server `10.0.1.2`. After that, restart the networking service (`sudo systemctl restart networking`)
```
auto enp0s31f6.4001
iface enp0s31f6.4001 inet static
  address 10.0.1.2
  netmask 255.255.255.0
  vlan-raw-device enp0s31f6
  mtu 1400
  up ip route add 10.0.0.0/16 via 10.0.1.1 dev enp0s31f6.4001
  down ip route del 10.0.0.0/16 via 10.0.1.1 dev enp0s31f6.4001
```
Detailed instructions can be found in Hetzner docs (https://docs.hetzner.com/cloud/networks/connect-dedi-vswitch/)

3. Create a new project in Hetzner Cloud Console and setup a new API token.
4. Manually create a new network with IP range `10.0.0.0/16`, name it `cluster-network` for example.
5. In this network, create an additional subnet `10.0.1.0/24` and connect it to the vSwitch. The default subnet `10.0.0.0/24` connects the cloud servers.
6. Create a cluster-config.yaml according to hetzner-k3s docs and make sure you use the network created earlier. 
```yaml
  private_network:
    enabled: true
    existing_network_name: "cluster-network"
```
Create the cluster running `hetzner-k3s create --config ./cluster-config.yaml`.
7. When everything is up and running, edit the `hcloud` secret and add `robot-user` and `robot-password`.
8. Edit the HCCM deployment and add the following environment variables
```yaml
        - name: HCLOUD_NETWORK_ROUTES_ENABLED
          value: "false"
        - name: ROBOT_ENABLED
          value: "true"
```

9. On the robot server, create the file `/etc/k8s-resolv.conf` with the single line `nameserver 8.8.8.8`.
10. Download Vito's installer script from https://gist.github.com/vitobotta/68b13edfe242fbf5c229bb347d7b69f6 and edit `SUBNET` and `K3S_TOKEN`. You can find the token in `/var/lib/rancher/k3s/server/token` on a master node.
11. Run the script. As @vitobotta already stated, it will hang in the end but it is ok to just cancel it.
12. You should now see the new node when running `kubectl get nodes -o wide`.
13. The new root node has an uninitialized taint that can simply be removed by running
```shell
kubectl taint nodes <node> node.cloudprovider.kubernetes.io/uninitialized-
```

Some additional notes:
  - If you use a loadbalancer, you have to manually add the root servers internal ip in Hetzner Cloud Console.
  - When HCCM restarts, it tries to spin up a HCCM pod on the root node which fails, but as far as I noticed, that does not really matter. All deployments I tested so far work fine and the node doesn't get tainted.

**Reply by vitobotta** [reply]
@if-mar Would you mind opening a PR with any additions etc to the guide I wrote that might be useful? Thanks!

**Reply by stremlau** [reply]
Starting with HCCM >= 1.24.0, it is sufficient to set HCLOUD_NETWORK_ROUTES_ENABLED=false, ROBOT_ENABLED=true and include the hcloud-user and password in the hcloud secret. No issues with the uninitialized taint.

**Comment by data-trm** [comment]
Hi @vitobotta, @if-mar 

Are there any updates on adding support for dedicated servers in hetzner-k3s?

We really enjoy using hetzner-k3s, but we’ve run into a limitation: the cloud servers no longer provide enough storage for our needs, and network volumes are too slow for our database workloads when hosted on these instances.

Because of this, we’re considering moving to dedicated servers. Ideally, we’d like to include these dedicated machines as part of our existing Kubernetes cluster, but we’re not entirely sure about the best way to approach this.

I came across this thread and plan to give it a try, but I wanted to check if this is still the recommended (or only) solution.

Thanks in advance for your help—and apologies if this isn’t the right place to ask!

**Reply by vitobotta** [reply]
Hi @data-trm - not yet unfortunately. I need to do this as soon as I can because we also need it at work for the same reason (storage). I have a lot on my plate at work plus other things and this task requires proper investigation so it needs some time.
